### PR TITLE
[Draft] `multipleOf` for some integers

### PR DESCRIPTION
**Pull request by hudson-ai**
TODO:
- [ ] expand to some decimals like `.1, .25, .001, ...`
- [ ] integrate with JSON

### CODE REVIEW

**Review by mmoskal — COMMENTED** [review]
looks great!

With an automaton, you can do it for any number N, but the automaton would have N states and we can't express automatons in regexes anyways. I like this approach more.

File: parser/src/json/numeric/multiple.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    ]))
+}
+
+#[cfg(test)]
+mod tests {
+    use derivre::RegexBuilder;
+    use super::rx_int_multiple_of;
+
+    #[test]
+    fn test_rx_int_multiple_of() {
```
[the remark this answers is not in the document]
Comment (hudson-ai) [reply parser/src/json/numeric/multiple.rs:106]: > With an automaton, you can do it for any number N, but the automaton would have N states and we can't express automatons in regexes anyways.
If you read the [derivation for multiples of 3](https://www.quaxio.com/triple/), they did indeed take the DFA approach, which is pretty nifty. They then rewrote the DFA in BNF and rewrote the rules until they got a regex. We could in principle do the same thing for other numbers. IMO, my approach is simple enough and works for a decently large set of numbers, so we can keep it until we get enough requests for other numbers, at which point we can rethink :)

> maybe also add a few tests for "multiple of not supported"
Yep yep :)

**Comment by mmoskal** [comment]
```
(llg) [15:06 json_stats]$ grep multipleOf tmp/num_files_with_feature.json 
  "multipleOf": 61,
  "multipleOf:0.01": 22,
  "multipleOf:1": 17,
  "multipleOf:1.0": 9,
  "multipleOf:0.1": 4,
  "multipleOf:16": 2,
  "multipleOf:0.25": 2,
  "multipleOf:12": 1,
  "multipleOf:256": 1,
  "multipleOf:3": 1,
  "multipleOf:2": 1,
  "multipleOf:10": 1,
  "multipleOf:4": 1,
```

This includes synthesized and handwritten tests, counts out of 9601

**Comment by mmoskal** [comment]
sorry, I meant to say this **excludes** the synthesized and handwritten tests (which use weird multiples)

**Comment by mmoskal** [comment]
Actually, I think this can be done in general with a derivative. We can introduce operator `rem(A,B)` meaning that the reminder of division by A is B. Now:

```
nullable(rem(_,B)) == true iff B == 0
deriv(K, rem(A,B)) = rem(A, (B*10 + K) % A) if byte K is a digit and no-match otherwise
```

**Comment by hudson-ai** [comment]
Just convincing myself...

Define

$$
\text{rem}(A, B) \equiv  \\{C \mid C \text{ mod } A = B \\}\.
$$

Let $C \in \text{rem}(A, B)$. Then $C = kA + B$ for some integer $k$.

Then for any integer $X$ (and in particular any $X \in \[0, …, 9\]$)

$$
10C + X = 10kA + 10B + X\,
$$

and therefore

$$
10C + X \in  \text{rem}(A, (10B + X)  \text{ mod } A)\.
$$

Super super slick @mmoskal. Color me impressed.

**Comment by mmoskal** [comment]
I should have believed @veanes when he told me everything that can be done with automata can be done with derivatives, but better :)

**Comment by mmoskal** [comment]
operator implemented in https://github.com/microsoft/derivre/pull/2

Note that you will likely want to use ReminderOf(m, m) to exclude empty string (which would normally be included in ReminderOf(m, 0))

**Comment by hudson-ai** [comment]
Closing in anticipation of better approach in derivre

**Pull request closed without merging**